modules/matchRoutes.js

```diff
@@ -99,25 +99,40 @@ function matchRouteDeep(
         params: createParams(paramNames, paramValues)
       }
 
-      getIndexRoute(route, location, function (error, indexRoute) {
+      if (typeof route.condition !== 'function') {
```

Comment:
I'd factor out the callback to `getIndexRoute` into a variable in this block, and just do something like:

```js
if (!route.condition) {
  getIndexRoute(route, location, finishMatch)
}
```

Or perhaps a better name instead of `finishMatch`.